### PR TITLE
 Implement AMD leaf 0x8000_001E (Processor Topology Information)

### PR DESCRIPTION
**Pull request by Umio-Yasuno**
Related: #62

It seems that the result of the comment part of the `ryzen_matisse` test file and the CPUID value are different.  
Do you want to leave it as is?

```
///   extended APIC ID = 10
///   Core Identifiers (0x8000001e/ebx):
///      core ID          = 0x5 (5)
///      threads per core = 0x2 (2)
///   Node Identifiers (0x8000001e/ecx):
///      node ID             = 0x0 (0)
///      nodes per processor = 0x1 (1)
```
```
    0x8000001e_00000000u64 => CpuIdResult { eax: 0x00000000, ebx: 0x00000100, ecx: 0x00000000, edx: 0x00000000 },
```

### CODE REVIEW

**Comment by gz** [comment]
> It seems that the result of the comment part of the ryzen_matisse test file and the CPUID value are different.

yea we should probably adjust the comment here. the reason there is this discrepancy is because I ran cpuid twice (once to get the raw output and once to get the description for the comment, and core ID returns different results based on wherever the process happens to run)

**Review by gz — COMMENTED** [review]
Thanks a lot this looks good!

One minor thing is to also extend the Intel machine tests with a trivial test that make sure that the `get_processor_topology_info` call returns None

File: src/extended.rs [hunk 1]
```diff
@@ -1635,3 +1507,193 @@ bitflags! {
...
+        }
+    }
+
+    /// x2APIC ID
+    pub fn x2apic_id(&self) -> u32 {
+        self.eax
+    }
+
+    /// Core ID
+    pub fn core_id(&self) -> u8 {
```
Comment (gz) [inline src/extended.rs:1541]: Maybe it might be better to stick to the AMD manual terminology here and use `compute_unit_id` instead of `core_id` 

From the manual:
"Compute unit ID. Identifies a Compute Unit, which may be one or more
physical cores that each implement one or more logical processors."

Core is a bit overloaded (never clear if hyperthreading or no?)
If we add this comment from the AMD manual to the documentation that should make things clearer too

File: src/extended.rs [hunk 2]
```diff
@@ -1635,3 +1507,193 @@ bitflags! {
...
+        get_bits(self.ebx, 0, 7) as u8
+    }
+
+    /// Threads per core
+    pub fn threads_per_core(&self) -> u8 {
```
Comment (gz) [inline src/extended.rs:1546]: `threads_per_compute_unit`?

File: src/extended.rs (hunk 1 above)
Comment (Umio-Yasuno) [reply src/extended.rs:1541]: AMD's documentation seems to use `Compute Unit` for family 15h-16h, and `Thread` for family 17h.  
I'll add that to the comments.  

BIOS and Kernel Developer’s Guide (BKDG) for AMD Family 15h Models 00h-0Fh Processors  
Page 501: <https://www.amd.com/system/files/TechDocs/42301_15h_Mod_00h-0Fh_BKDG.pdf>

Processor Programming Reference (PPR) for AMD Family 17h Model 01h, Revision B1 Processors  
Page 82: <https://developer.amd.com/wordpress/media/2017/11/54945_PPR_Family_17h_Models_00h-0Fh.pdf>

**Comment by gz** [comment]
Great thanks!

**Comment by Umio-Yasuno** [comment]
Thanks!
